src/Service/Resolvers/MsSqlQueryBuilder.cs

```diff
@@ -107,31 +110,72 @@ public string Build(SqlExecuteStructure structure)
+                $"{queryToGetCountOfRecordWithPK};";
+
+            // Final query to be executed for the given PUT/PATCH operation.
+            StringBuilder upsertQuery = new(prefixQuery);
+
+            // Query to update record (if there exists one for given PK).
+            StringBuilder updateQuery = new(
+                $"IF @ROWS_TO_UPDATE = 1" +
+                $"UPDATE {tableName} WITH(UPDLOCK) " +
+                $"SET {updateOperations} " +
+                $"OUTPUT {outputColumns} " +
+                $"WHERE {updatePredicates};");
+
+            // Append the update query to upsert query.
+            upsertQuery.Append(updateQuery);
+
+            if (!structure.IsFallbackToUpdate)
             {
-                return $"SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;BEGIN TRANSACTION; UPDATE {QuoteIdentifier(structure.DatabaseObject.SchemaName)}.{QuoteIdentifier(structure.DatabaseObject.Name)} " +
-                    $"WITH(UPDLOCK) SET {Build(structure.UpdateOperations, ", ")} " +
-                    $"OUTPUT {MakeOutputColumns(structure.OutputColumns, OutputQualifier.Inserted)} " +
-                    $"WHERE {predicates} " +
-                    $"IF @@ROWCOUNT = 0 " +
-                    $"BEGIN; " +
-                    $"INSERT INTO {QuoteIdentifier(structure.DatabaseObject.SchemaName)}.{QuoteIdentifier(structure.DatabaseObject.Name)} ({Build(structure.InsertColumns)}) " +
-                    $"OUTPUT {MakeOutputColumns(structure.OutputColumns, OutputQualifier.Inserted)} " +
-                    $"VALUES ({string.Join(", ", structure.Values)}) " +
-                    $"END; COMMIT TRANSACTION";
+                // Append the conditional to check if the insert query is to be executed or not.
+                // Insert is only attempted when no record exists corresponding to given PK.
+                upsertQuery.Append("ELSE ");
+
+                // Columns which are assigned some value in the PUT/PATCH request.
+                string insertColumns = Build(structure.InsertColumns);
+
+                // Predicates added by virtue of database policy for create operation.
+                string createPredicates = JoinPredicateStrings(structure.DbPolicyPredicatesForOperations[Config.Operation.Create]);
+
+                // Query to insert record (if there exists none for given PK).
+                StringBuilder insertQuery = new($"INSERT INTO {tableName} ({insertColumns}) OUTPUT {outputColumns}");
+
+                string fetchColumnValuesQuery = BASE_PREDICATE.Equals(createPredicates) ?
+                    $"VALUES({string.Join(", ", structure.Values)});" :
+                    $"SELECT {insertColumns} FROM (VALUES({string.Join(", ", structure.Values)})) T({insertColumns}) WHERE {createPredicates};";
+
+                // Append the values to be inserted to the insertQuery.
+                insertQuery.Append(fetchColumnValuesQuery);
```

Comment:
Can we reuse the function `Build(SqlInsertQueryStructure)` to build this insertQuery?

Reply:
We could place the common code into another method but are not really going to benifit from it. Its simpler as it is right now.

Reply:
Ok